src/plugins/section_category.ts

```diff
@@ -0,0 +1,257 @@
+      currCategory === category._id
+    )
+    menu.appendChild(item)
+  })
+
+  document.addEventListener('mousedown', (event) => {
+    if (!menu.contains(event.target as Node)) {
+      popper.destroy()
+    }
+  })
+
+  return menu
+}
+
+function createButton(
+  view: EditorView,
+  props: SectionCategoryProps,
+  node: ManuscriptNode,
+  pos: number,
+  category: string,
+  canEdit = true
+) {
+  const arrow = document.createElement('div')
+  arrow.className = 'section-category popper-arrow'
+  const button = document.createElement('button')
+  button.innerHTML = sectionCategoryIcon
+  button.className = `section-category-button ${category && 'assigned'}`
+  if (canEdit) {
+    button.addEventListener('mousedown', (e) => {
+      e.stopPropagation()
+      popper.destroy()
+      const menu = createMenu(view, props, node, pos, category)
+      popper.show(button, menu, 'bottom-end', false, [
+        { name: 'offset', options: { offset: [0, -10] } },
+      ])
+    })
+  } else {
+    button.addEventListener('mouseenter', () => {
+      const tooltip = document.createElement('div')
+      tooltip.className = 'section-category tooltip'
+      tooltip.textContent = 'Category:'
+      const span = document.createElement('span')
+      span.textContent = getCategoryName(props.sectionCategories, category)
+      tooltip.appendChild(span)
+      tooltip.appendChild(arrow)
+      popper.show(button, tooltip, 'left', false, [
+        { name: 'offset', options: { offset: [0, 10] } },
+        { name: 'arrow', options: { element: arrow } },
+      ])
+    })
+    button.addEventListener('mouseleave', () => {
+      popper.destroy()
+    })
+  }
+
+  return button
+}
+
+function buildPluginState(
+  state: EditorState,
+  props: SectionCategoryProps
+): PluginState {
+  const decorations: Decoration[] = []
+  const can = props.getCapabilities()
+
+  state.doc.descendants((node, pos) => {
+    if (
+      node.type.name === 'section' &&
+      isEditableSectionCategory(node.attrs.category) &&
+      !isUnique(node.attrs.category)
+    ) {
+      decorations.push(
+        Decoration.widget(pos + 1, (view) =>
+          createButton(
+            view,
+            props,
+            node,
+            pos,
+            node.attrs.category,
+            can?.editArticle
+          )
+        )
+      )
+    }
+  })
+
+  return { decorations: DecorationSet.create(state.doc, decorations) }
+}
+
+function getExistingCatsCounted(state: EditorState): Record<string, number> {
+  const existingCatsCounted: Record<string, number> = {}
+  state.doc.descendants((node) => {
+    if (
+      node.type.name === 'section' &&
+      node.attrs.category?.startsWith('MPSectionCategory:')
+    ) {
+      existingCatsCounted[node.attrs.category] =
+        (existingCatsCounted[node.attrs.category] || 0) + 1
+    }
+  })
+  return existingCatsCounted
+}
+
+function getSortedSectionCategories(
+  state: EditorState,
+  container: ManuscriptNode,
```

Comment:
Probably would be better to type it as SectionNode if I'm not missing anything.

Reply:
done